File: backend/app/schemas/market_signal.py
from pydantic import BaseModel, ConfigDict
from datetime import date, datetime
from typing import Any


class MacroLayer(BaseModel):
    cycle_phase: str | None = None
    gdp_trend: str | None = None
    inflation_level: str | None = None
    liquidity: str | None = None
    interest_rate: str | None = None
    score: float | None = None


class GeoLayer(BaseModel):
    overall_risk: float | None = None
    risk_level: str | None = None
    safe_haven_demand: str | None = None
    score: float | None = None


class IndustryLayer(BaseModel):
    heatmap: dict[str, float] | None = None
    recommended: list[str] | None = None
    avoid: list[str] | None = None
    score: float | None = None


class SocialLayer(BaseModel):
    major_themes: list[str] | None = None
    theme_strength: dict[str, float] | None = None
    consumer_confidence: str | None = None
    score: float | None = None


class InternalLayer(BaseModel):
    equity_bond_spread: float | None = None
    sentiment: str | None = None
    style_rotation: str | None = None
    volatility_regime: str | None = None
    score: float | None = None


class MarketSignalRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    date: date
    macro_cycle_phase: str | None
    macro_score: float | None
    geo_overall_risk: float | None
    geo_score: float | None
    industry_heatmap: dict[str, Any] | None
    industry_score: float | None
    social_major_themes: list[str] | None
    social_score: float | None
    internal_sentiment: str | None
    internal_score: float | None
    composite_score: float | None
    market_mood: str | None
    market_cycle: str | None
    created_at: datetime


class MarketSignalLatest(BaseModel):
    date: date
    composite_score: float
    market_mood: str
    market_cycle: str
    macro: MacroLayer
    geo: GeoLayer
    industry: IndustryLayer
    social: SocialLayer
    internal: InternalLayer

File: frontend/src/api/marketSignal.ts
import { delay } from './mock/utils'
import { DEMO_MARKET_SIGNAL } from './mock/demoData'

export interface MacroLayer {
  cycle_phase: string | null
  gdp_trend: string | null
  inflation_level: string | null
  liquidity: string | null
  interest_rate: string | null
  score: number | null
  cycle_analysis?: CycleAnalysis | null
}

export interface CycleAnalysis {
  final_phase: string
  final_description: string
  final_asset_preference: string
  confidence: number
  fused_coordinates: Record<string, number>
  model_results: CycleModelResult[]
  consistency: number
  data_completeness: number
}

export interface CycleModelResult {
  model: string
  phase: string
  description: string
  asset_preference: string
  score: number
  inputs: Record<string, number | null>
}

export interface GeoLayer {
  overall_risk: number | null
  risk_level: string | null
  safe_haven_demand: string | null
  score: number | null
}

export interface IndustryLayer {
  heatmap: Record<string, number> | null
  recommended: string[] | null
  avoid: string[] | null
  score: number | null
}

export interface SocialLayer {
  major_themes: string[] | null
  theme_strength: Record<string, number> | null
  consumer_confidence: string | null
  score: number | null
}

export interface InternalLayer {
  equity_bond_spread: number | null
  sentiment: string | null
  style_rotation: string | null
  volatility_regime: string | null
  score: number | null
}

export interface MarketSignalLatest {
  date: string
  composite_score: number
  market_mood: string
  market_cycle: string
  cycle_analysis?: CycleAnalysis | null
  macro: MacroLayer
  geo: GeoLayer
  industry: IndustryLayer
  social: SocialLayer
  internal: InternalLayer
}

export const marketSignalApi = {
  getLatest: () => delay(400).then(() => DEMO_MARKET_SIGNAL as MarketSignalLatest),
  collect: () => delay(500).then(() => ({ success: true })),
}
